src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/MessageInterception/DefaultInterceptorManager.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+#nullable enable
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage.MessageInterception
+{
+    [Export(typeof(InterceptorManager))]
+    internal sealed class DefaultInterceptorManager : InterceptorManager
+    {
+        private readonly IList<Lazy<MessageInterceptor, IInterceptMethodMetadata>> _lazyInterceptors;
+
+        [ImportingConstructor]
+        public DefaultInterceptorManager([ImportMany] IEnumerable<Lazy<MessageInterceptor, IInterceptMethodMetadata>> lazyInterceptors)
+        {
+            _ = lazyInterceptors ?? throw new ArgumentNullException(nameof(lazyInterceptors));
+            _lazyInterceptors = lazyInterceptors.ToList();
+        }
+
+        public override bool HasInterceptor(string messageName)
+        {
+            foreach (var interceptor in _lazyInterceptors)
+            {
+                foreach (var method in interceptor.Metadata.InterceptMethods)
+                {
+                    if (method.Equals(messageName, StringComparison.Ordinal))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public override async Task<JToken?> ProcessInterceptorsAsync(string methodName, JToken token, string sourceLanguageName, CancellationToken cancellationToken)
+        {
+            for (var i = 0; i < _lazyInterceptors.Count; i++)
+            {
+                var item = _lazyInterceptors[i];
+                if (CanInterceptMessage(methodName, item.Metadata.InterceptMethods))
+                {
+                    var result = await item.Value.ApplyChangesAsync(token, sourceLanguageName, cancellationToken);
+                    if (result.UpdatedToken is null)
+                    {
+                        // The interceptor has blocked this message
+                        return null;
+                    }
+
```

Comment:
Probably worth doing a cts.throw if requested between each apply to avoid doing excess work.